File: merv/src/merv/brain/workflows/tools.py
"""MCP tool contracts for the workflow engine.

Workflows owns the schema of every ``workflow.*`` call; the support registry
in ``surface/tools/contracts.py`` merges this table with the other owners'.
"""

from __future__ import annotations

from typing import Any

from pydantic import Field, model_validator

from ..kernel.tools import ContractModel, ProjectScopedInput, ToolContract


class WorkflowStatusAndNextInput(ProjectScopedInput):
    instance_id: str | None = Field(default=None, description="Any workflow instance id (lens, synthesis, plugin workflow).")
    experiment_id: str | None = None
    task_id: str | None = None

    @model_validator(mode="after")
    def _one_scope(self) -> "WorkflowStatusAndNextInput":
        if sum(bool(value) for value in (self.instance_id, self.experiment_id, self.task_id)) > 1:
            raise ValueError("pass only one of instance_id, experiment_id or task_id")
        return self


class WorkflowInstanceInput(ProjectScopedInput):
    instance_id: str = Field(min_length=1)


class WorkflowBeginInput(WorkflowInstanceInput):
    expected_revision: int = Field(ge=0, description="The revision workflow.status_and_next returned.")


class WorkflowStartInput(ProjectScopedInput):
    workflow: str = Field(min_length=1)
    request_id: str = Field(min_length=1, description="Stable id for this start; reuse it on retries.")
    data: dict[str, Any] = Field(default_factory=dict)
    entry: str = ""
    version: int | None = Field(default=None, ge=1)


class WorkflowTransitionInput(WorkflowInstanceInput):
    action: str = Field(min_length=1)
    expected_revision: int = Field(ge=0, description="The revision workflow.status_and_next returned.")
    request_id: str = Field(min_length=1, description="Stable id for this action; reuse it on retries.")
    payload: dict[str, Any] = Field(default_factory=dict)


TOOLS: dict[str, ToolContract] = {
    "workflow.status_and_next": ToolContract(
        handler_identity="application.status_for_agent",
        input_model=WorkflowStatusAndNextInput,
        description=(
            "Start or resume here. Unscoped: the project's next step and context (introduction, literature, claims, "
            "one row per experiment and task). With experiment_id or task_id: that record's state, revision, next "
            "action, blockers and documents. With instance_id: any other instance. Then call the tool the next action names."
        ),
    ),
    "workflow.catalog": ToolContract(
        handler_identity="workflows.catalog", input_model=ContractModel, visibility="internal",
        description="List registered workflows and their definition versions.",
    ),
    "workflow.start": ToolContract(
        handler_identity="workflows.start", input_model=WorkflowStartInput, visibility="internal",
        description="Runner effect: start a registered workflow at an entry, pinned to a version. Agents create records with experiment.create and task.create.",
    ),
    "workflow.transition": ToolContract(
        handler_identity="workflows.transition", input_model=WorkflowTransitionInput,
        description="Apply a graph action to an instance that has no dedicated tool (a reflection lens, project synthesis, a plugin workflow); experiments and tasks use experiment.transition and task.transition. Pass the revision from workflow.status_and_next.",
    ),
    "workflow.assignment": ToolContract(
        handler_identity="workflows.assignment", input_model=WorkflowInstanceInput,
        description="The current node's brief, input references and handoff instructions; unmet prerequisites are listed instead.",
    ),
    "workflow.begin": ToolContract(
        handler_identity="workflows.begin", input_model=WorkflowBeginInput,
        description="Interactive sessions call this once before working a node: it records the start at the given revision and returns the assignment. Auto-run leases begin on their own.",
    ),
    "workflow.history": ToolContract(
        handler_identity="workflows.history", input_model=WorkflowInstanceInput,
        description="State history, delivery status and pending repairs for one instance.",
    ),
}

File: merv-typescript/packages/workflows/src/tools.ts
import type { Context } from 'cordis';
import type {} from '@merv/api/types';
import type { Caller, Data, WorkflowBegin, WorkflowExtendLimit } from '@merv/contracts';
import { z } from 'zod';

/** Generic node guidance and activation; domain tools own the work's handoff. */
export const workflowToolsPlugin = {
  name: 'merv-workflow-tools',
  inject: ['workflows', 'tools'],
  apply(ctx: Context) {
    const workflows = ctx.workflows;
    ctx.effect(() =>
      ctx.tools.register({
        name: 'workflow.status_and_next',
        description:
          'Start or resume here. With instanceId (the task ID for a task), read the current gate, caller-specific next action, blockers, references and revision. limits reports each loop limit leaving the current state; at gate loop_limit_reached every allowed return is used and the work waits for a human. Omit instanceId for a project overview of all workflow instances, where such work is listed under escalated. Optionally preflight an action with proposed input; this does not execute it or reserve permission. Commands recheck current rules. Use stable requestId values when calling mutation tools.',
        readOnly: true,
        inputSchema: z
          .object({
            instanceId: z.string().min(1).optional(),
            action: z.string().min(1).optional(),
            input: z.record(z.unknown()).optional(),
          })
          .strict()
          .superRefine((input, context) => {
            if (input.action !== undefined && input.instanceId === undefined)
              context.addIssue({
                code: z.ZodIssueCode.custom,
                message: 'action requires instanceId',
              });
            if (input.input !== undefined && input.action === undefined)
              context.addIssue({ code: z.ZodIssueCode.custom, message: 'input requires action' });
          }),
        handler: async (
          caller: Caller,
          input: { instanceId?: string; action?: string; input?: Data },
        ) =>
          input.instanceId
            ? await workflows.evaluate(caller, input.instanceId, {
                action: input.action,
                input: input.input,
              })
            : await workflows.overview(caller),
      }),
    );
    ctx.effect(() =>
      ctx.tools.register({
        name: 'workflow.catalog',
        description:
          "Read the deployed workflow definitions: each program's states, its initial and terminal states, and the actions that connect them. This is the shape of the machine, never any record's place in it.",
        readOnly: true,
        inputSchema: z.object({}).strict(),
        handler: async () => workflows.catalog(),
      }),
    );
    ctx.effect(() =>
      ctx.tools.register({
        name: 'workflow.assignment',
        description:
          'Read your current node assignment: brief, context, evidence references, declared tools and handoff. Assignment guards are checked without recording a work start or saving a context package. An eligible reviewer can inspect an open review assignment; claim it through review.start before saving review context or submitting a verdict.',
        readOnly: true,
        inputSchema: z.object({ instanceId: z.string().min(1) }).strict(),
        handler: async (caller: Caller, input: { instanceId: string }) =>
          await workflows.assignment(caller, input.instanceId),
      }),
    );
    ctx.effect(() =>
      ctx.tools.register({
        name: 'workflow.process',
        description:
          "Read one instance's process graph, derived on read: the pinned definition's states and edges, which edges the record shows were actually taken and by whom, the live status and blockers of the current state's outgoing edges, and dependency edges to other instances. An edge means the machinery stepped through a gate; it never says a finding is correct. Per-record detail stays with the domain read.",
        readOnly: true,
        inputSchema: z.object({ instanceId: z.string().min(1) }).strict(),
        handler: async (caller: Caller, input: { instanceId: string }) =>
          await workflows.process(caller, input.instanceId),
      }),
    );
    ctx.effect(() =>
      ctx.tools.register({
        name: 'workflow.begin',
        description:
          'Before beginning interactive node work, pass its instanceId and current expectedRevision. Rechecks your assignment and records its first start once per revision, then returns the assignment. Retries preserve the first start without changing workflow state or revision. No requestId is needed. This does not create a worker lease or claim a review.',
        readOnly: false,
        inputSchema: z
          .object({
            instanceId: z.string().min(1),
            expectedRevision: z.number().int().nonnegative(),
          })
          .strict(),
        handler: async (caller: Caller, input: WorkflowBegin) =>
          await workflows.begin(caller, input),
      }),
    );
    ctx.effect(() =>
      ctx.tools.register({
        name: 'workflow.extend_limit',
        conversation: 'propose',
        description:
          'Allow one workflow instance more rounds of a loop limit it has exhausted (gate loop_limit_reached). Project admins only, never a leased worker. Pass the limit name from status_and_next limits, how many additional returns to allow, a reason and a stable requestId. The grant is recorded and adds to earlier grants; the instance keeps its state and revision. Each additional return buys one more automated review.',
        readOnly: false,
        inputSchema: z
          .object({
            instanceId: z.string().min(1),
            limit: z.string().min(1),
            additional: z.number().int().min(1).max(100),
            reason: z.string().trim().min(1).max(500),
            requestId: z.string().min(1).max(256),
          })
          .strict(),
        handler: async (caller: Caller, input: WorkflowExtendLimit) =>
          await workflows.extendLimit(caller, input),
      }),
    );
  },
};
export default workflowToolsPlugin;
